packages-exp/auth-exp/src/core/user/token_manager.ts

```diff
@@ -76,6 +69,17 @@ export class StsTokenManager {
   async getToken(
     auth: AuthInternal,
     forceRefresh = false
+  ): Promise<string | null> {
+    if (!this.isPassthroughMode) {
+      return this.getTokenAndRefreshIfNeeded(auth, forceRefresh);
+    } else {
+      return this.getTokenAndTriggerCallback(/* auth, forceRefresh */);
+    }
+  }
+
+  private async getTokenAndRefreshIfNeeded(
+    auth: AuthInternal,
+    forceRefresh = false
   ): Promise<string | null> {
     _assert(
       !this.accessToken || this.refreshToken,
```

Comment:
Hmm good q good q, taking a look at where `getToken()` is called - it's only called in `UserImpl.getIdToken()` and the latter case would still be caught by an assert (see [this line](https://github.com/firebase/firebase-js-sdk/blob/master/packages-exp/auth-exp/src/core/user/user_impl.ts#L99)), though the user would not be logged out. I don't have enough context around if this is the expected behavior for when both the access token and refresh token are not present though - perhaps @avolkovi / @sam-gc can weigh in?

As for the impact on how the `refreshToken` check for passthrough happens, could you expand a bit more on this? The only case in which we'd hit the incorrect branching logic in `getToken()` is if a call to `getToken()` happens prior to a call to `updateFromServerResponse()`, but at least for `signInWithCustomToken()`, this would not be the case. But please let me know if there are any blindspots I'm missing!

Reply:
This code is mirroring this bit in the original SDK:
https://github.com/firebase/firebase-js-sdk/blob/a69220b519185239c364f7a5b3d5983e9c87d7c1/packages/auth/src/token.js#L279-L299

I'm not entirely sure _why_ we need to check for the presence of `accessToken` as well, but we just copied that bit over. In this case though it's fine, as this function is the one that gets called when we're _not_ in passthrough mode. In this code, `getTokenAndRefreshIfNeeded()` is called for non-passthrough, while `getTokenAndTriggerCallback()` is the equivalent (unimplemented so far) for passthrough mode. (@lisajian please correct me if I'm wrong)

Reply:
Cool, thanks for the link to the OG expiration check! And yep ditto^^, that all checks out on my end